File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.java
class Solution {
    public int[] validSequence(String word1, String word2) {
        int n = word1.length();
        int m = word2.length();
        int[] last = new int[m];
        int j = m - 1;
        for (int i = n - 1; i >= 0; i--) {
            if (j >= 0 && word1.charAt(i) == word2.charAt(j)) {
                last[j] = i;
                j--;
            }
        }
        int[] ans = new int[m];
        j = 0;
        int usedMismatch = 0;
        for (int i = 0; i < n && j < m; i++) {
            if (word1.charAt(i) == word2.charAt(j)) {
                ans[j] = i;
                j++;
            }
            else if (usedMismatch == 0 &&
                    (j == m - 1 || i + 1 <= last[j + 1])) {
                ans[j] = i;
                j++;
                usedMismatch = 1;
            }
        }
        if (j != m) {
            return new int[0];
        }
        return ans;
    }
}

File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.py
class Solution:
    def validSequence(self, word1: str, word2: str) -> List[int]:
        n = len(word2)
        last = [-1]*n
        j = n-1
        for i, ch in reversed(list(enumerate(word1))): 
            if j >= 0 and ch == word2[j]: 
                last[j] = i 
                j -= 1
        j = cnt = 0 
        ans = []
        for i, ch in enumerate(word1): 
            if j < n: 
                if ch == word2[j] or cnt == 0 and (j == n-1 or i+1 <= last[j+1]): 
                    if ch != word2[j]: cnt = 1
                    ans.append(i)
                    j += 1
        return ans if j == n else []